### PR TITLE
Avoid Host & Device Sync In LR Scheduler

### LINKED ISSUE #133662: Host & Device Sync In LR Scheduler
### 🐛 Describe the bug

We observed performance regression with the most recent PyTorch, and finally noticed a new host & device synchronizaiton introduced in LR scheduler at here: https://github.com/pytorch/pytorch/blob/main/torch/optim/lr_scheduler.py#L250.

The `lr.item()` is where the sync invoked:

```python
for i, data in enumerate(zip(self.optimizer.param_groups, values)):
    param_group, lr = data
    if isinstance(param_group["lr"], Tensor):
        lr_val = lr.item() if isinstance(lr, Tensor) else lr  # type: ignore[attr-defined]
        param_group["lr"].fill_(lr_val)
    else:
        param_group["lr"] = lr
```

This is origianlly introduced at here: https://github.com/pytorch/pytorch/pull/123753.

### Versions

```text
Collecting environment information...
PyTorch version: 2.5.0a0+872d972e41.nvInternal
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Mar 22 2024, 16:50:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-70-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version: 550.54.14
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.3.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Silver 4314 CPU @ 2.40GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        6
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4800.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       768 KiB (16 instances)
L1i cache:                       512 KiB (16 instances)
L2 cache:                        20 MiB (16 instances)
L3 cache:                        24 MiB (1 instance)
NUMA node(s):                    2
NUMA node0 CPU(s):               0-7,16-23
NUMA node1 CPU(s):               8-15,24-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Vulnerable
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] onnx==1.16.0
[pip3] open-clip-torch==2.20.0
[pip3] optree==0.12.1
[pip3] pytorch-lightning==2.4.0
[pip3] pytorch-triton==3.0.0+dedb7bdf3
[pip3] torch==2.5.0a0+872d972e41.nvinternal
[pip3] torch-tensorrt==2.5.0a0
[pip3] torchdiffeq==0.2.4
[pip3] torchmetrics==1.4.0.post0
[pip3] torchsde==0.2.6
[pip3] torchvision==0.20.0a0
[conda] Could not collect
```

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by alpha0422**
Fixes #133662.

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
Thanks! Good call, I think this was left over after some refactoring.

We can actually keep fill_, as my timing experiment shows it's faster in both the Tensor (c) and python value (b) case.

image

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -247,8 +247,7 @@ def step(self, epoch: Optional[int] = None):
         for i, data in enumerate(zip(self.optimizer.param_groups, values)):
             param_group, lr = data
             if isinstance(param_group["lr"], Tensor):
-                lr_val = lr.item() if isinstance(lr, Tensor) else lr  # type: ignore[attr-defined]
-                param_group["lr"].fill_(lr_val)
+                param_group["lr"].copy_(lr)
```
Comment (janeyx99) [inline torch/optim/lr_scheduler.py:250]:
```suggestion
                param_group["lr"].fill_(lr)
```

**Comment by alpha0422** [comment]
> Thanks! Good call, I think this was left over after some refactoring.
> 
> We can actually keep fill_, as my timing experiment shows it's faster in both the Tensor (c) and python value (b) case.
> 
> image

Thanks, that makes sense to me.

**Comment by janeyx99** [comment]
Once you sign the CLA we can merge the PR!

**Review by janeyx99 — APPROVED** [review]

**Review by eqy — APPROVED** [review]

**Comment by alpha0422** [comment]
> Once you sign the CLA we can merge the PR!

Thanks! The process is in progress and it will be finished soon.

**Comment by alpha0422** [comment]
@janeyx99 I've got the authorization, please proceed the flow, thanks!

**Pull request closed without merging**